Pipeline/Segmentation/SegmSubmodules/Evaluation.py

```diff
@@ -16,48 +16,46 @@ def draw_mask_plots(prediction, ground_truth, plot_file):
-    """Counts metrics for passed prediction and saves it to file in .JSON format. Supported metrics:
+def count_metrics_on_sample(prediction, ground_truth, json_file, metrics_list=None):
+    """Counts metrics for passed prediction and saves it to file in .JSON format. Default metrics:
     Pearson's correlation coefficient, f1-score, log_loss, roc_auc, precision, recall, intersection over union
 
     Args:
         prediction: the prediction binary mask vector
         ground_truth: the ground truth binary mask vector
         json_file: path to json file to save metrics
-        metrics_dict: dictionary with custom metrics names as keys and metrics functions as values,
-        metric function has 2 arguments: ground_truth and prediction"""
+        metrics_list: list with custom metrics,
+        metric function has 2 arguments: ground_truth and prediction. If None, default metrics used"""
+    if metrics_list is None:  # assign default metrics
+        metrics_list = [f1_score, log_loss, roc_auc_score, precision_score, recall_score, __intersection_over_union,
+                        __corr]
+
     with open(json_file, 'w') as f:
         json_dict = {}  # init dictionary to save
 
-        corr = np.corrcoef(ground_truth, prediction)[1, 0]  # get Pearson's correlation coefficient
-        json_dict["corr"] = corr
-
-        f1 = f1_score(ground_truth, prediction)
-        json_dict["f1"] = float(f1)
-
-        log_loss_val = log_loss(ground_truth, prediction)
-        json_dict["log_loss"] = float(log_loss_val)
+        for metric in metrics_list:  # init
+            json_dict[metric.__name__] = 0
 
-        roc_auc = roc_auc_score(ground_truth, prediction)
-        json_dict["roc_auc"] = float(roc_auc)
+        sample_cnt = prediction.shape[0]
+        divisor = 0  # for normalization
+        for current_num in range(sample_cnt):
+            if K.sum(ground_truth[current_num,:] * prediction[current_num,:]) > 0:  # don't evaluate sample without song
+                divisor += 1
+                for metric in metrics_list:
+                    json_dict[metric.__name__] += float(metric(ground_truth[current_num,:],
+                                                               prediction[current_num,:]))
 
-        precision = precision_score(ground_truth, prediction)
-        json_dict["precision"] = float(precision)
-
-        recall = recall_score(ground_truth, prediction)
-        json_dict["recall"] = float(recall)
-
-        iou = __intersection_over_union(ground_truth, prediction)
-        json_dict["IoU"] = float(iou)
-
-        if metrics_dict is not None:  # add custom metrics
-            for metric_name in metrics_dict.keys():
-                metric_val = metrics_dict[metric_name](ground_truth, prediction)
-                json_dict[metric_name] = float(metric_val)
+        for metric in metrics_list:  # normalize
+            json_dict[metric.__name__] /= divisor
```

Comment:
you can divide directly while adding - (1 + 1) / 2 = 1/2 + 1/2

moreover, please, check if all the metrics are additive - I mean if it is correct that you use the mean value on subsets

Reply:
There are 2 reasons to divide in the end:

Reply:
Agree!